httpx/private_ip_validator.go

```diff
@@ -80,29 +80,54 @@ var _ http.RoundTripper = (*NoInternalIPRoundTripper)(nil)
 
 // NoInternalIPRoundTripper is a RoundTripper that disallows internal IP addresses.
 type NoInternalIPRoundTripper struct {
-	http.RoundTripper
 	internalIPExceptions []string
 }
 
 func (n NoInternalIPRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
-	rt := http.DefaultTransport
-	if n.RoundTripper != nil {
-		rt = n.RoundTripper
-	}
-
 	incoming := IncomingRequestURL(request)
 	incoming.RawQuery = ""
 	incoming.RawFragment = ""
 	for _, exception := range n.internalIPExceptions {
 		if incoming.String() == exception {
-			return rt.RoundTrip(request)
+			return http.DefaultTransport.RoundTrip(request)
 		}
 	}
 
-	host, _, _ := net.SplitHostPort(request.Host)
-	if err := DisallowIPPrivateAddresses(stringsx.Coalesce(host, request.Host)); err != nil {
-		return nil, err
-	}
+	return NoInternalTransport.RoundTrip(request)
+}
+
+var NoInternalDialer = &net.Dialer{
+	Timeout:   30 * time.Second,
+	KeepAlive: 30 * time.Second,
+	Control: func(network, address string, _ syscall.RawConn) error {
+		if !(network == "tcp4" || network == "tcp6") {
```

Comment:
From docs:

````
	// If Control is not nil, it is called after creating the network
	// connection but before actually dialing.
	//
	// Network and address parameters passed to Control method are not
	// necessarily the ones passed to Dial. For example, passing "tcp" to Dial
	// will cause the Control function to be called with "tcp4" or "tcp6".
```

Not sure if that means that the address could also be different?

Another question is: how does this deal with sockets? They should also not be allowed.

Reply:
The `address` will contain the resolved address, so if you `Dial("tcp", "google.com")`, the `Control callback will receive the resolved IP address, and most likely `tcp6` as the protocol. We check that we get only `tcp4` or `tcp6` and that the address is really an IP address (we don't do any lookup any more, and err if it is not a parseable IP address).

Regarding your other question, I expanded the test coverage, see: https://github.com/ory/x/commit/33e13f6b3a01b9b6b88c6348909bddd251d65c39#diff-ece808bf9f537051020f575d16079a3f53528a677d892509b133015eed4741dfR87-R125. In short, only TCP traffic is allowed, nothing else.